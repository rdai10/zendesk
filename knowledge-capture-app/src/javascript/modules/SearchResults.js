import React from 'react';
import { Col, Grid, Row } from '@zendeskgarden/react-grid';
import { Well, Title, Paragraph } from '@zendeskgarden/react-notifications';
import { Ellipsis, MD, SM } from '@zendeskgarden/react-typography';

import I18n from '../lib/i18n';
import { displayDateInMDYFormat } from '../lib/utility';

export default function SearchResults({ results }) {
	return (
		<Grid>
			{results.map((result) => (
				<Result key={result.id} result={result} />
			))}
		</Grid>
	);
}

function Result({ result }) {
	return (
		<Row>
			<Col>
				<Well>
					<Title>
						<MD isBold>{result.name}</MD>
					</Title>

					<SM>
						<Ellipsis>
							{/* <Paragraph size="small">breadcrumb</Paragraph> */}
						</Ellipsis>

						<Paragraph size="small">
							{I18n.t('last edited')}{' '}
							{displayDateInMDYFormat(result.edited_at)}
						</Paragraph>
					</SM>
				</Well>
			</Col>
		</Row>
	);
}
