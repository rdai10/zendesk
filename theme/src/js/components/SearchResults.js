import preact from 'preact';
import PropTypes from 'prop-types';

import 'core-js/fn/array/includes';

import {
	getArticlesBySearch,
	getSectionBySectionId
} from '../helpers/api-helpers';

import LoadingIndicator from './LoadingIndicator';
import Pagination from './Pagination';
import SearchFilter from './SearchFilter';

const ARTICLES_PER_PAGE = 10;

class SearchResultBreadCrumb extends preact.Component {
	constructor(props) {
		super(props);

		this.state = {
			breadcrumb: []
		};
	}

	componentDidMount() {
		const {id, locale} = this.props;

		getSectionBySectionId(id, locale, 'categories')
			.then(
				({data}) => {
					const [categories] = data.categories;

					const breadcrumbData = [
						{
							name: categories.name,
							url: categories.html_url
						},
						{
							name: data.section.name,
							url: data.section.html_url
						}
					];

					this.setState(
						{
							breadcrumb: breadcrumbData
						}
					);
				}
			)
			.catch(
				(err) => {
					if (process.env.NODE_ENV === 'development') {
						console.log(err);
					}
				}
			);
	}

	render() {
		const {breadcrumb} = this.state;

		return (
			<div>
				{!!breadcrumb.length && (
					<ol class="breadcrumbs">
						{breadcrumb.map(
							(data, index) => (
								<li key={index} title={data.name}>
									<a href={data.url}>{data.name}</a>
								</li>
							)
						)}
					</ol>
				)}
			</div>
		);
	}
}

SearchResultBreadCrumb.PropTypes = {
	id: PropTypes.string.isRequired,
	locale: PropTypes.string.isRequired
};

class SearchResults extends preact.Component {
	constructor(props) {
		super(props);

		this.displayNoResultsMsg = this.displayNoResultsMsg.bind(this);
		this.handlePaginationClick = this.handlePaginationClick.bind(this);
		this.handleSearchFilterChange = this.handleSearchFilterChange.bind(this);
		this.querySearchResults = this.querySearchResults.bind(this);
		this.updateResultsCount = this.updateResultsCount.bind(this);

		this.state = {
			loading: true,
			productLabel: '',
			results: [],
			totalPage: 0
		};
	}

	componentDidMount() {
		this.querySearchResults();
	}

	displayNoResultsMsg(bool) {
		const noResults = document.getElementById(
			'noResults'
		);

		if (bool) {
			noResults.classList.add('show');
		}
		else {
			noResults.classList.remove('show');
		}
	}

	handlePaginationClick(currentPage) {
		const {locale, queryString} = this.props;
		const {productLabel} = this.state;

		getArticlesBySearch(
			queryString,
			ARTICLES_PER_PAGE,
			currentPage,
			productLabel,
			locale
		)
			.then(
				({data}) => {
					this.setState(
						{
							results: data.results
						}
					);
				}
			)
			.catch(
				(err) => {
					if (process.env.NODE_ENV === 'development') {
						console.log(err);
					}

					this.setState(
						{
							loading: true
						}
					);
				}
			);

		window.scroll(0, 0);
	}

	handleSearchFilterChange(label) {
		this.setState(
			{
				loading: true,
				productLabel: label
			}
		);

		this.querySearchResults(label);
	}

	querySearchResults(label) {
		const {locale, queryString} = this.props;

		if (queryString) {
			getArticlesBySearch(queryString, ARTICLES_PER_PAGE, 1, label, locale)
				.then(
					({data}) => {
						this.setState(
							{
								loading: false,
								results: data.results,
								totalPage: data.page_count
							}
						);

						this.displayNoResultsMsg(
							!data.results.length
						);

						this.updateResultsCount(data.count);
					}
				)
				.catch(
					(err) => {
						if (process.env.NODE_ENV === 'development') {
							console.log(err);
						}
					}
				);
		}
		else {
			this.setState(
				{
					loading: false
				}
			);

			this.displayNoResultsMsg(true);
			this.updateResultsCount(0);
		}
	}

	updateResultsCount(count) {
		const searchResultsCount = document.getElementById('searchResultsCount');

		searchResultsCount.innerHTML = count;
	}

	render(
		{fastTrackLabel, filterLabel, filterOptions, locale},
		{loading, results, totalPage}
	) {
		return (
			<div>
				<SearchFilter
					label={filterLabel}
					onChange={this.handleSearchFilterChange}
					options={filterOptions}
				/>

				{!loading && !!results.length && (
					<ul class="search-results-list">
						{results.map(
							(result) => (
								<li key={result.id} class="search-result" id={result.id}>
									<div class="search-results-header">
										<a class="search-result-link semibold" href={result.html_url}>{result.title}</a>

										{result.label_names.includes(
											'Fast Track'
										) && (
											<span class="initiative-label label label-sm">
												{fastTrackLabel}
											</span>
										)}
									</div>

									<h5 class="search-result-description"
										dangerouslySetInnerHTML={{__html: result.snippet}} />

									<SearchResultBreadCrumb
										id={result.section_id}
										locale={locale}
									/>
								</li>
							)
						)}
					</ul>
				)}

				{!loading && totalPage > 1 && (
					<Pagination
						onClick={this.handlePaginationClick}
						total={totalPage}
					/>
				)}

				{loading && <LoadingIndicator />}
			</div>
		);
	}
}

SearchResults.PropTypes = {
	fastTrackLabel: PropTypes.string.isRequired,
	filterLabel: PropTypes.string.isRequired,
	filterOptions: PropTypes.arrayOf(
		PropTypes.shape(
			{
				displayName: PropTypes.string,
				value: PropTypes.string
			}
		)
	).isRequired,
	locale: PropTypes.string.isRequired,
	queryString: PropTypes.string.isRequired
};

export default SearchResults;