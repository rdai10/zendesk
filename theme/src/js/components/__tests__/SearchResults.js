import preact from 'preact';
import {cleanup, render} from 'preact-testing-library';

import SearchResults from '../SearchResults';

jest.mock('../../helpers/api-helpers');

afterEach(cleanup);

describe('SearchResults', () => {
	const filterOptions = [
		{displayName: 'Option 1', value: 'one'},
		{displayName: 'Option 2', value: 'two'}
	];

	it('renders correctly', () => {
		const {container} = render(
			<SearchResults
				filterLabel='filter'
				filterOptions={filterOptions}
				locale='en-us'
				queryString='Test'
			/>
		);

		expect(container).toBeTruthy();
	});
});
