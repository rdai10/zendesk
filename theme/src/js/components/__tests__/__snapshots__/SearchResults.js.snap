// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults renders correctly 1`] = `
<div>
  <div>
    <div
      class="search-filter"
    >
      <label
        class="semi-bold"
        for="productSearchFilter"
      >
        filter
      </label>
      <select
        class="col-md-3 input-small"
        id="productSearchFilter"
      >
        <option
          value="one"
        >
          Option 1
        </option>
        <option
          value="two"
        >
          Option 2
        </option>
      </select>
    </div>
    
    <span
      aria-hidden="true"
      class="loading-animation loading-animation-sm"
    />
  </div>
</div>
`;
